File: covid.py
# list of countries, country has object containing, name, susceptible pop., infected pop., immune pop.
# simulate spread (travel btwn countries?)
import math
import matplotlib.pyplot as plt
from matplotlib.ticker import MultipleLocator

# 
class country():
    def __init__(self, country, suscept, infect, immune):
        self.name = country
        self.suscept = suscept
        self.infect = infect
        self.immune = immune

# 85k current cases, 83.5k recoveries, 6.5k deaths countrywide as of march 15
# assume the dead are immune
infect = 2
recovd = 0
populn = 7800000
# per day multiplication of cases
factor = 1.10999
# infection period in days
days = 43
# simulation length in days
length = 66

# initialize country
world = country("world", populn-infect-recovd, infect, recovd)

fortnight = [-1] * round(days)

def spread(factor, country, days, list):
    if country.suscept == 0 or country.infect == 0: # none left to infect or eradicated
        return
    
    # number of new cases with logistic growth
    infected = (factor-1)*country.infect*(1-((country.infect+country.immune)/country.suscept))
    country.suscept -= infected # subtract new cases from suscept

    if country.suscept < 0: # prevent suscept from going negative
        country.infect += country.suscept
        country.suscept = 0
        return
    
    country.infect += infected # add new cases to infected

    for i in range(1, round(days)):
        list[round(days)-i] = list[round(days)-i-1]
        list[0] = country.infect

def recover(factor, country, days, list):
    if country.infect == 0: # if eradicated
        return
    
    # number of recoveries
    if list[round(days)-1] == -1:
        recovered = country.infect/(math.pow(factor, days)) # approximation of cases that were in progress days ago # subtract recovered from infected
    else: 
        recovered = list[round(days)-1]
    
    country.infect -= recovered

    if country.infect < 0: # prevent infect from going negative
        country.immune += country.infect
        country.infect = 0
        return

    country.immune += recovered # add recovered to immune

# aggregate infections list
aggrlist = [None] * (length)

# simulate over the length
for i in range(1, length+1):
    spread(factor, world, days, fortnight)
    recover(factor, world, days, fortnight)
    aggrlist[i-1] = world.infect+world.immune # add cumulative to list
    # print simulation results by day
    print("Day: {:d} Susceptible: {:.3f} Infected: {:.3f} Immune: {:.3f} Cumulative: {:.3f} ".format(i, world.suscept, world.infect, world.immune, world.infect+world.immune))

# plot simulation results
#plt.yticks(ticks=range(0,1000000,100000))
plt.grid(b=True)
plt.plot(range(1,length+1), aggrlist)
plt.savefig("aggr.png")

File: covid.c
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#define DAYS 14

// C implementation of covid.py, used for regression on to data points to get parameters
// Mostly because this is faster than the python implementation

typedef struct country {
    char country[10];
    double suscept;
    float infect;
    float immune;
} country;

int *fortnight;

void spread(float factor, country *world, float days, int *arr) {
    if(world->suscept == 0 || world->infect == 0) {
        return;
    }
    float infected = (factor-1)*world->infect*(1-((world->infect+world->immune)/world->suscept));
    world->suscept -= infected;
    if(world->suscept < 0) {
        world->infect += world->suscept;
        world->suscept = 0;
        return;
    }
    world->infect += infected;

    int rdays = (int) nearbyintf(days);
    for(int i = 1; i < rdays; i++) {
        arr[rdays-i] = arr[rdays-i-1];
        arr[0] = world->infect;
    }
}

void recover(float factor, country *world, float days, int *arr) {
    if(world->suscept == 0) {
        return;
    }
    float recovered;
    int rdays = (int) nearbyintf(days);
    if(arr[rdays-1] == -1) {
        recovered = world->infect/(pow(factor, rdays));
    } else {
        recovered = arr[rdays-1];
    }

    world->infect -= recovered;
    if(world->infect < 0) {
        world->immune += world->infect;
        world->infect = 0;
        return;
    }
    world->immune += recovered;
}

// Implement a regression algorithm to get growth factor and
// infectious period parameters from start and end data points.

typedef struct result {
    float factor;
    float days;
} result;

result regression() {
    float initfactor = 1.3;
    float initdays = 5;
    struct country world;
    strcpy(world.country, "world");
    // regression initial conditions for start
    int populn = 7800000; // world pop.
    float factor = initfactor; // init growth factor
    float days = initdays; // init infectious period
    int infect = 2; // init infections
    int recovd = 0; // init recoveries
    int length = 66; // length of each simulation in days

    // regression targets
    float infend = 634;
    float immend = 224;

    // margin between result and target
    float margin = 100;
    // step of parameters after each loop
    float step = 0.000001;

    while(1) {
        // init world struct
        world.infect =  (float) infect;
        world.immune = (float) recovd;
        world.suscept = (double) populn-infect-recovd;
        int rdays = (int) nearbyintf(days);
        fortnight = malloc(sizeof(int)*rdays);
        for(int i = 0; i < rdays; i++) {
            fortnight[i] = -1;
        } 

        // complete a simulation over the length
        for(int i = 0; i < length; i++) {
            spread(factor, &world, days, fortnight);
            recover(factor, &world, days, fortnight);
        }

        // difference between targets and results
        float infdiffer = infend-world.infect;
        float immdiffer = immend-world.immune;

        // adjust parameters accordingly if outside margin
        if(world.infect <= infend-margin || world.infect >= infend+margin) {
            factor += step*infdiffer;
        }
        if(world.immune >= immend+margin || world.immune <= immend-margin) {
            days += step*immdiffer;
            if(days < 1) {
                days = initdays;
            }
        }
        // break if parameters all within margin
        if(world.infect <= infend+margin && world.infect >= infend-margin && world.immune <= immend+margin && world.immune >= immend-margin) {
            break;
        }
        free(fortnight);
    }

    // return regression results
    result ret;
    ret.factor = factor;
    ret.days = days;
    return ret;
}


int main() {
    // do the regression
    result ret;
    ret = regression();
    printf("factor %.6f days %.6f\n", ret.factor, ret.days);

    // // mostly same as in covid.py
    // int infect = 85;
    // int recovd = 90;
    // int populn = 7800000;
    // float factor = 1.189060;
    // int days = 14.451526;
    // int length = DAYS;
    // int rdays = (int) nearbyintf(days);
    // fortnight = malloc(sizeof(int)*rdays);
    // for(int i = 0; i < rdays; i++) {
    //     fortnight[i] = -1;
    // } 

    // country world;
    // strcpy(world.country, "world");
    // world.suscept = (float) populn-infect-recovd;
    // world.infect =  (float) infect;
    // world.immune = (float) recovd;

    // int aggr[DAYS]; // DAYS defined up top
    // for(int i = 1; i < length+1; i++) {
    //     spread(factor, &world, days, fortnight);
    //     recover(factor, &world, days, fortnight);
    //     aggr[i] = world.infect+world.immune;
    //     printf("Day: %d Susceptible: %.3f Infected: %.3f Immune: %.4f Cumulative: %.3f\n", i, world.suscept, world.infect, world.immune, world.infect+world.immune);
    // }
    // free(fortnight);
    return 0;
}
